conn.go

```
@@ -299,316 +293,281 @@ func (c *conn) initTLSConfig() {
+
+	// drain pending RX error
+	var rxErr error
+	select {
+	case rxErr = <-c.rxErr:
+		// there was an error in connReader
+		if errors.Is(rxErr, net.ErrClosed) {
+			// this is the expected error when the connection is closed, swallow it
+			rxErr = nil
+		}
+	default:
+		// no pending read error
+	}
+
+	if txErr == nil && rxErr == nil && closeErr == nil {
+		// if there are no errors, it means user initiated close() and we shut down cleanly
+		c.doneErr = &ConnectionError{}
+	} else if amqpErr, ok := rxErr.(*Error); ok {
+		// we experienced a peer-initiated close that contained an Error.  return it
+		c.doneErr = &ConnectionError{inner: amqpErr}
+	} else if txErr != nil {
+		c.doneErr = &ConnectionError{inner: txErr}
+	} else if rxErr != nil {
+		c.doneErr = &ConnectionError{inner: rxErr}
+	} else {
+		c.doneErr = &ConnectionError{inner: closeErr}
+	}
 }
 
 // Err returns the connection's error state after it's been closed.
 // Calling this on an open connection will block until the connection is closed.
 func (c *conn) Err() error {
-	c.errMu.Lock()
-	defer c.errMu.Unlock()
-	return &ConnectionError{inner: c.err}
+	<-c.Done
+	return c.doneErr
 }
 
-// mux is started in it's own goroutine after initial connection establishment.
-// It handles muxing of sessions, keepalives, and connection errors.
-func (c *conn) mux() {
-	var (
-		// allocated channels
-		channels = bitmap.New(uint32(c.channelMax))
+func (c *conn) NewSession() (*Session, error) {
+	c.sessionsByChannelMu.Lock()
+	defer c.sessionsByChannelMu.Unlock()
 
-		// create the next session to allocate
-		// note that channel always start at 0, and 0 is special and can't be deleted
-		nextChannel, _ = channels.Next()
-		nextSession    = newSessionResp{session: newSession(c, uint16(nextChannel))}
+	// create the next session to allocate
+	// note that channel always start at 0
+	channel, ok := c.channels.Next()
+	if !ok {
+		return nil, fmt.Errorf("reached connection channel max (%d)", c.channelMax)
```

Nice to not ignore an error, unlikely as it might be!